lib/classes/Variables.js

```diff
@@ -0,0 +1,178 @@
+const replaceall = require('replaceall');
+
+class Variables {
+
+  constructor(serverless) {
+    this.serverless = serverless;
+    this.service = this.serverless.service;
+
+    this.variableSyntax = RegExp(this.service.defaults.variableSyntax, 'g');
+    this.overwriteSyntax = RegExp(/,/g);
+    this.fileRefSyntax = RegExp(/^file\(([a-zA-Z0-9._\-\/]+?)\)/g);
+    this.envRefSyntax = RegExp(/^env:./g);
+    this.optRefSyntax = RegExp(/^opt:./g);
+    this.selfRefSyntax = RegExp(/^self:./g);
+  }
+
+  populateService(processedOptions) {
+    const that = this;
+    this.options = processedOptions || {};
+    const variableSyntaxProperty = this.service.defaults.variableSyntax;
+
+    // temporally remove variable syntax from service otherwise it'll match
+    this.service.defaults.variableSyntax = true;
+
+    /*
+     * we can't use an arrow function in this case cause that would
+     * change the lexical scoping required by the traverse module
+     */
+    traverse(this.service).forEach(function (propertyParam) {
+      const t = this;
+      let property = propertyParam;
+
+      if (typeof property === 'string') {
+        property = that.populateProperty(property);
+        t.update(property);
+      }
+    });
+
+    this.service.defaults.variableSyntax = variableSyntaxProperty;
+    return this.service;
+  }
+
+  populateProperty(propertyParam) {
+    let property = propertyParam;
+    let valueToPopulate;
+    if (typeof property === 'string' && property.match(this.variableSyntax)) {
+      property.match(this.variableSyntax).forEach((matchedString) => {
+        const variableString = matchedString
+          .replace(this.variableSyntax, (match, varName) => varName.trim())
+          .replace(/\s/g, '');
+
+        if (variableString.match(this.overwriteSyntax)) {
+          valueToPopulate = this.overwrite(variableString);
+        } else {
+          valueToPopulate = this.getValueFromSource(variableString);
+        }
+
+        property = this.populateVariable(property, matchedString, valueToPopulate);
+      });
+      return this.populateProperty(property);
+    }
+    return property;
+  }
+
+  populateVariable(propertyParam, matchedString, valueToPopulate) {
+    let property = propertyParam;
+
+    if (typeof valueToPopulate === 'string') {
+      property = replaceall(matchedString, valueToPopulate, property);
+    } else {
+      if (property !== matchedString) {
+        const errorMessage = [
+          'Trying to populate non string value into',
+          ` a string for variable ${matchedString}.`,
+          ' Please make sure the value of the property is a string.',
+        ].join('');
+        throw new this.serverless.classes
+          .Error(errorMessage);
+      }
+      property = valueToPopulate;
+    }
+    return property;
+  }
+
+  overwrite(variableStringsString) {
+    const variableStringsArray = variableStringsString.split(',');
+    const variablesValues = variableStringsArray
```

Comment:
This implementation means we evaluate every variable. This an be a problem e.g. when we have a file syntax that shouldn't be evaluated, e.g.

`runtime: ${self:custom.production.runtime, self:custom.dev.runtime, file(./testfile123123)}`

```
custom:
  stage: production
  dev:
    runtime: JAVA9
```

`self:custom.dev.runtime` can be evaluated, but it still fails the whole command as the file can't be found:

```
ENOENT: no such file or directory, open '/app/tmp/serverless-integration-test-aws-nodejs/testfile123123
```

Please only evaluate if necessary